### PR TITLE
Build docker container in GitHub actions

### PR DESCRIPTION
**Pull request by obbardc**
This is a continuation of work done by @eds-collabora in !275

This replaces the old, simpler pipeline with a three phase process:

- First, build the image and cache it using docker buildx.
- Second, run all the tests in parallel, restoring the image from the cache.
- Thirdly, if the tests pass:
  - if this is a push to the main branch, push to DockerHub.
  - push to GitHub Container registry (PS: will push to a user's own fork).

This uses Buildkit caching aggressively, and will make use of the entire 5GiB allocation of cache space that GitHub provides over time.

It requires the following additional repository secrets:

- DOCKERHUB_USERNAME: the username to login as on DockerHub (e.g. go-debos)
- DOCKERHUB_PASSWORD: an access token for the DockerHub repository.

Closes: #275

### CODE REVIEW

**Comment by obbardc** [comment]
@sjoerdsimons I think this is the most important thing to get right next, since Dockerhub builds are disabled we currently are not be able to build any new images.

File: .github/workflows/build-test.yaml [hunk 1]
```diff
@@ -1,29 +1,246 @@
...
+      - name: Checkout repository
+        uses: actions/checkout@v2
+
+      - name: Extract Docker metadata
+        id: meta
+        uses: docker/metadata-action@v3
+        with:
+          images: ${{ env.GITHUB_TAG }}
+          tags: |
+            "type=raw,value=latest,enable=${{ github.ref == format('refs/heads/{0}', github.event.repository.default_branch) }}"
```
Comment (sjoerdsimons) [inline .github/workflows/build-test.yaml:159]: you probably want to set:
```
flavor: |
  latest = auto
```
to avoid generating latest on tagging?

Comment (obbardc) [reply .github/workflows/build-test.yaml:159]: from what i have seen `docker/metadata-action` is quite fickle and only creates the latest tag for releases which are tagged in git which doesn't work for our workflow; we want any push to the default branch to also be tagged as latest.

I'll do some more testing on a fork though :-)

Comment (sjoerdsimons) [reply .github/workflows/build-test.yaml:159]: sorry i meant set `latest = false`

But we should really move to pushing :latest on tags/releases indeed rather then on every push

Comment (obbardc) [reply .github/workflows/build-test.yaml:159]: right that's a great way forward then - I will remove this hack then ;-)

File: .github/workflows/build-test.yaml [hunk 2]
```diff
@@ -1,29 +1,246 @@
...
+          context: .
+          push: false
+          tags: debos
+          file: docker/Dockerfile
+          cache-from: type=local,src=/tmp/.build-cache
+          load: true
 
       - name: Run test recipes using UML backend
         run: |
           docker-compose -f docker/recipes-test-uml.yml \
```
Comment (sjoerdsimons) [inline .github/workflows/build-test.yaml:134]: now that no builds are done in docker-hub anymore can we get rid of the docker-compose wrapping? (can be later though)

Comment (obbardc) [reply .github/workflows/build-test.yaml:134]: yeah - i also want to move Dockerfile to the root and combine the readme files - added a TODO :-)

**Comment by sjoerdsimons** [comment]
@obbardc  can you rebas ethis; seems it's now impossible to merge

**Pull request merged**